via-router/src/router.rs

```diff
@@ -109,75 +108,76 @@ impl<T> Router<T> {
     }
 
     pub fn visit<'a>(&'a self, path: &str) -> Vec<Binding<'a, T>> {
-        // The following three lines allocate unconditionally.
-        let mut results = Vec::with_capacity(8);
-        let mut first = Vec::with_capacity(1);
-        let segments = path::split(path);
+        let mut segments = Split::new(path).lookahead();
 
-        let mut offset = 0;
-        let mut queue = SmallVec::<[usize; 6]>::new();
-        let mut next = SmallVec::<[usize; 6]>::new();
+        let mut bindings = Vec::with_capacity(8);
+        let mut branch = Vec::with_capacity(64);
+        let mut nodes = Vec::with_capacity(1);
+        let mut next = SmallVec::<[&[usize]; 2]>::new();
 
         if let Some(root) = self.tree.first() {
-            queue.extend_from_slice(&root.children);
-            first.push(Match::new(segments.is_empty(), None, &root.route));
-            results.push(Binding::new(None, first));
+            nodes.push(Match::new(!segments.has_next(), None, &root.route));
+            bindings.push(Binding::new(None, nodes));
+            branch.extend_from_slice(&root.children);
         }
 
-        loop {
-            let mut nodes = Vec::new();
-            let (range, exact) = match segments.get(offset) {
-                Some(to) => {
-                    offset += 1;
-                    (*to, offset == segments.len())
-                }
-                None => {
-                    for key in queue.drain(..) {
-                        let node = lookup!(&self.tree, key);
-                        let route = &node.route;
-
-                        if let Pattern::Wildcard(param) = &node.pattern {
-                            nodes.push(Match::new(true, Some(param), route));
-                        }
-                    }
-
-                    results.push(Binding::new(None, nodes));
-                    return results;
-                }
-            };
+        while let Some((range, has_next)) = segments.next() {
+            let mut nodes = Vec::with_capacity(4);
```

Comment:
I plan on coming up with some sort of data structure to get rid of the allocation per path segment that we are doing currently. 

A self-referential struct where binding holds indices to the matched nodes is what I'm thinking. Reducing allocations and increased cache locality is all I would expect from such a refactor.

Reply:
I'm thinking this would be a great addition to a follow up PR.